### PR TITLE
Custom attribute ui

### LINKED ISSUE #120: Unknown attribute JSON should be a custom interact-able component and not a raw string

### PR DESCRIPTION
**Pull request by tlein**
Resolves #120

### CODE REVIEW

**Comment by tlein** [comment]
Should we add an add button for arrays?

File: src/duckling/controls/json.component.ts [hunk 1]
```diff
@@ -36,29 +79,48 @@ export class JsonComponent implements OnChanges {
...
+        for (let key in this.value) {
+            keys.push(key);
         }
+        return keys;
     }
 
-    onValueChanged(newValue : string) {
-        this._displayedValue = JSON.parse(newValue);
-        this.valueChanged.emit(this._displayedValue);
+    noop(element : any, i : number) {
```
Comment (jeffswenson) [inline src/duckling/controls/json.component.ts]: I don't see where this is used.

File: src/duckling/controls/section.component.scss [hunk 2]
```diff
@@ -2,6 +2,10 @@
 @import '../colors';
 @import '../forms';
 
+:host {
```
Comment (jeffswenson) [inline src/duckling/controls/section.component.scss:5]: It seems like a bad idea to style the host. What makes this necessary?

File: src/duckling/controls/section.component.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ import {
             </dk-section-header>
             <dk-section-header
                 *ngIf="!collapsible"
+                [collapsible]="false"
```
Comment (jeffswenson) [inline src/duckling/controls/section.component.ts:23]: You don't need to specify this. You made it the default.

File: src/duckling/forms.scss [hunk 4]
```diff
@@ -6,12 +6,18 @@ $form-label-padding: $form-label-padding-top $form-label-padding-right $form-lab
 
 $spacer-unit: 13px;
 
-.form-label {
+.form-label:not(:first-child) {
```
Comment (jeffswenson) [inline src/duckling/forms.scss]: You could dry this up by nesting the first-child selector within form-label and overriding the padding there.

File: src/duckling/controls/json.component.ts [hunk 5]
```diff
@@ -8,24 +8,67 @@ import {
...
+    styleUrls: ['./duckling/controls/json.component.css'],
     template: `
-        <dk-validated-input
-            label="JSON"
-            [value]="valueAsJSON"
-            [validator]="isValidJSON"
-            (validInput)="onValueChanged($event)">
-        </dk-validated-input>
+        <div *ngFor="let key of jsonKeys">
+            <dk-number-input
```
Comment (jeffswenson) [inline src/duckling/controls/json.component.ts:32]: The intent would be more obvious if this was an ngSwitch

**Review by jeffswenson — CHANGES_REQUESTED** [review]
Overall I think it looks really good.  Few nit picks.

File: src/duckling/controls/section.component.scss (hunk 2 above)
Comment (tlein) [reply src/duckling/controls/section.component.scss:5]: Necessary for margin-top to take effect

File: src/duckling/controls/section.component.ts (hunk 3 above)
Comment (tlein) [reply src/duckling/controls/section.component.ts:23]: flipped the default to true so now I do need to specify this

File: src/duckling/controls/json.component.ts (hunk 1 above)
Comment (tlein) [reply src/duckling/controls/json.component.ts]: removed, it was test code

File: src/duckling/forms.scss (hunk 4 above)
Comment (tlein) [reply src/duckling/forms.scss]: fixed

File: src/duckling/controls/json.component.ts (hunk 5 above)
Comment (tlein) [reply src/duckling/controls/json.component.ts:32]: fixed

**Review by jeffswenson — APPROVED** [review]

**Pull request merged**